code/components/gta-net-five/src/netTimeSync.cpp

```diff
@@ -0,0 +1,338 @@
+			m_replySequence = reqSequence;
+
+			// progressive backoff once we've established a valid time base
+			if (m_effectiveTimeBetweenSyncs < m_configMaxBackoff)
+			{
+				m_effectiveTimeBetweenSyncs = std::min(m_configMaxBackoff, m_effectiveTimeBetweenSyncs * 2);
+			}
+
+			m_retryCount = 0;
+
+			// use flag 4 to reset time at least once, even if game session code has done so to a higher value
+			if (!(m_applyFlags & 4))
+			{
+				m_lastTime = m_timeDelta + timeGetTime();
+			}
+
+			m_applyFlags |= 7;
+		}
+		else
+		{
+			m_nextSync = 0;
+			m_retryCount++;
+		}
+
+		// update average RTT
+		m_lastRtt = (rtt + m_lastRtt) / 2;
+	}
+	else
+	{
+		m_nextSync = 0;
+		m_retryCount++;
+	}
+}
+
+template<int Build>
+bool rage::netTimeSync<Build>::IsInitialized()
+{
+	if (!g_initedTimeSync)
+	{
+#ifdef IS_RDR3
+		// we don't want to use cloud time
+		m_useCloudTime = false;
+#endif
+
+		g_origInitializeTime(this, _getConnectionManager(), 1, nullptr, 0, nullptr, 7, 2000, 60000);
+
+		// to make the game not try to get time from us
+		m_connectionMgr = nullptr;
+
+		g_initedTimeSync = true;
+
+		return false;
+	}
+
+	return (m_applyFlags & 4) != 0;
+}
+
+template<int Build>
+inline void rage::netTimeSync<Build>::SetConnectionManager(netConnectionManager* mgr)
+{
+	m_connectionMgr = mgr;
+}
+
+bool sync::IsWaitingForTimeSync()
+{
+#ifdef GTA_FIVE
+	if (xbr::IsGameBuildOrGreater<2372>())
+	{
+		return !(*g_netTimeSync<2372>)->IsInitialized();
+	}
+	else if (xbr::IsGameBuildOrGreater<2060>())
+	{
+		return !(*g_netTimeSync<2060>)->IsInitialized();
+	}
+
+	return !(*g_netTimeSync<1604>)->IsInitialized();
+#elif IS_RDR3
+	if (xbr::IsGameBuildOrGreater<1355>())
+	{
+		return !(*g_netTimeSync<1355>)->IsInitialized();
+	}
+
+	return !(*g_netTimeSync<1311>)->IsInitialized();
+#endif
+}
+
+static inline void TimeSyncMainGameFrameUpdate()
+{
+#if GTA_FIVE
+	if (xbr::IsGameBuildOrGreater<2372>())
+	{
+		(*g_netTimeSync<2372>)->Update();
+	}
+	else if (xbr::IsGameBuildOrGreater<2060>())
+	{
+		(*g_netTimeSync<2060>)->Update();
+	}
+	else
+	{
+		(*g_netTimeSync<1604>)->Update();
+	}
+#elif IS_RDR3
+	if (xbr::IsGameBuildOrGreater<1355>())
+	{
+		(*g_netTimeSync<1355>)->Update();
+	}
+	else
+	{
+		(*g_netTimeSync<1311>)->Update();
+	}
+#endif
+}
+
+static inline void HandleTimeSyncUpdatePacket(net::Buffer& buf)
+{
+#ifdef GTA_FIVE
+	if (xbr::IsGameBuildOrGreater<2372>())
+	{
+		(*g_netTimeSync<2372>)->HandleTimeSync(buf);
+	}
+	else if (xbr::IsGameBuildOrGreater<2060>())
+	{
+		(*g_netTimeSync<2060>)->HandleTimeSync(buf);
+	}
+	else
+	{
+		(*g_netTimeSync<1604>)->HandleTimeSync(buf);
+	}
+#elif IS_RDR3
+	if (xbr::IsGameBuildOrGreater<1355>())
+	{
+		(*g_netTimeSync<1355>)->HandleTimeSync(buf);
+	}
+	else
+	{
+		(*g_netTimeSync<1311>)->HandleTimeSync(buf);
+	}
+#endif
+}
+
+static InitFunction initFunction([]()
+{
+	NetLibrary::OnNetLibraryCreate.Connect([](NetLibrary* lib)
+	{
+		lib->AddReliableHandler("msgTimeSync", [](const char* data, size_t len)
```

Comment:
What is your opinion on moving the `xbr::IsGameBuildOrGreater` checks outside of the `ReliableHandlerType` callback. For example:

```cpp
// Beneath SetConnectionManager?
template<int Build>
static void HandleTimeSyncUpdatePacket(const char* data, size_t len)
{
	net::Buffer buf(reinterpret_cast<const uint8_t*>(data), len);
	(*g_netTimeSync<Build>)->HandleTimeSync(buf);
}
```

Then having the `InitFunction` something like (with brackets?):

```cpp
static InitFunction initFunction([]()
{
	static constexpr char kMsgTimeSync[] = "msgTimeSync"; // Eventually moved?
	NetLibrary::OnNetLibraryCreate.Connect([](NetLibrary* lib)
	{
#ifdef GTA_FIVE
		if (xbr::IsGameBuildOrGreater<2372>())
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<2372>);
		else if (xbr::IsGameBuildOrGreater<2060>())
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<2060>);
		else
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<1604>);
#elif IS_RDR3
		if (xbr::IsGameBuildOrGreater<1355>())
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<1355>);
		else
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<1311>);
#endif
	});
});
```

Reply:
Is there any reason to do it now? This approach as well as the one I'm currently using isn't something that we should keep anyway. The initializer lists makes more sense but would require more planning.

Reply:
Agreed.